File: apps/0-1k/610/1.0/powershell/bin/searchsplunk.fs
// Copyright 2011 - Michael Vierling

(*
To use this module from Powershell, follow this example:

ps>Search-Splunk "My splunk search query" -Host "splunk.yellowpages.com" -Port "8089" -Option "&max_count=100"
*)

module SplunkSearch

open System
open System.Net
open System.Security
open System.Management.Automation

type TrustAll() =
    interface System.Net.ICertificatePolicy with
        member this.CheckValidationResult( p, s, w, i ) = true;;

[<Cmdlet(System.Management.Automation.VerbsCommon.Search,"Splunk")>]
type public MyWriteInputObjectCmdlet() =
    inherit Cmdlet()

    let mutable _Port = "8089"
    let mutable _Host = "splunk.yellowpages.com"
    let mutable _SearchString = ""
    let mutable _Credentials = new System.Net.NetworkCredential( "search_user", "password" )
    let mutable _Option = ""

    [<Parameter(Mandatory=true, Position = 0, ValueFromPipeline = true)>]
    member this.SearchString
        with get() = _SearchString
        and set(v) = _SearchString <- v
    [<Parameter(Mandatory=false)>]
    member this.Port
        with get() = _Port
        and set(v) = _Port <- v

    [<Parameter(Mandatory=false)>]
    member this.Host
        with get() = _Host
        and set(v) = _Host <- v

    [<Parameter(Mandatory=false)>]
    member this.Credentials
        with get() = _Credentials
        and set(v) = _Credentials <- v

    [<Parameter(Mandatory=false)>]
    member this.Option
        with get() = _Option
        and set(v) = _Option <- v

    override this.ProcessRecord() =
        let trust = new TrustAll()
        System.Net.ServicePointManager.set_CertificatePolicy( trust )
        let request = "https://" + this.Host + ":" + this.Port + "/services/search/jobs/export"
        let post = "?search=search " + this.SearchString + "&exec_mode=oneshot" + this.Option
        let request_post = request + post
        let splunk = new System.Net.WebClient()
        splunk.Credentials <- this.Credentials
        let page = splunk.DownloadString( request_post )
        this.WriteObject( page.ToString() )

File: apps/0-1k/610/1.0/powershell/bin/searchsplunk.cs
// Copyright 2011 Michael Vierling

using System;
using System.Collections.Generic;
using System.Text;
using System.Net;
using System.Net.Security;
using System.Security;
using System.Security.Cryptography.X509Certificates;
using System.Management.Automation;

namespace SearchSplunk
{
    [Cmdlet(System.Management.Automation.VerbsCommon.Search,"Splunk")]
    public class MySplunkCmdlet : Cmdlet
    {
        // callback used to validate the certificate in an SSL conversation
        private static bool ValidateRemoteCertificate(
            object sender,
            X509Certificate certificate,
            X509Chain chain,
            SslPolicyErrors policyErrors
        )
        {
            return true;
        }

        private string _Port = "8089";
        private string _Host = "splunk.yellowpages.com";
        private string _SearchString = "";
        private NetworkCredential _Credentials = new System.Net.NetworkCredential("search_user", "password");
        private string _Option = "";

        [Parameter(Mandatory = true, Position = 0, ValueFromPipeline = true)]
        public string SearchString
        {
            get { return _SearchString; }
            set { _SearchString = value; }
        }

        [Parameter(Mandatory = false)]
        public string Port
        {
            get { return _Port; }
            set { _Port = value; }
        }

        [Parameter(Mandatory = false)]
        public string Host
        {
            get { return _Host; }
            set { _Host = value; }
        }

        [Parameter(Mandatory = false)]
        public NetworkCredential Credentials
        {
            get { return _Credentials; }
            set { _Credentials = value; }
        }

        [Parameter(Mandatory = false)]
        public string Option
        {
            get { return _Option; }
            set { _Option = value; }
        }

        protected override void ProcessRecord()
        {
            System.Net.ServicePointManager.ServerCertificateValidationCallback +=
                new RemoteCertificateValidationCallback(ValidateRemoteCertificate);
            var request = "https://" + this.Host + ':' + this.Port + "/services/search/jobs/export";
            var post = "?search=search " + this.SearchString + "&exec_mode=oneshot" + this.Option;
            var request_post = request + post;
            var splunk = new System.Net.WebClient();
            splunk.Credentials = this.Credentials;
            var page = splunk.DownloadString( request_post );
            WriteObject(page);
        }
    }
}


